tests/unit/test_http_client.py

```diff
@@ -223,3 +223,87 @@ def simple_json_callback(resp):
     assert exp_resp == resp
     assert 1 == mock_set_valid_token_called
     assert 2 == len(caplog.records)
+
+
+#####
+# Tests & fixtures the PagingGoogleClientMixin
+#####
+@pytest.fixture
+def simple_paging_client(mocker, get_gce_client):
+    class TestClient(http_client.PagingGoogleClientMixin):
+        def parse_response_into_items(self, response, items):
+            items.append(response['data'])
+
+        async def get_json(self, url, params=None):
+            return {'data': 'data', 'nextPageToken': 'token'}
+
+    return TestClient()
+
+
+def test_parse_response_into_items_not_implemented():
+    class DummyClient(http_client.PagingGoogleClientMixin):
+        pass
+
+    paging_base = DummyClient()
+    with pytest.raises(NotImplementedError):
+        paging_base.parse_response_into_items(None, [])
+        paging_base._session.close()
+
+
+@pytest.mark.asyncio
+async def test_list_all_items_failure(
+        mocker, mock_coro, simple_paging_client, caplog):
+    """All requests fail so an empty list should be returned."""
+    caplog.set_level(logging.WARNING)
+    sleep_mock, sleep_coro = mock_coro
+    mocker.patch('gordon_janitor_gcp.http_client.asyncio.sleep', sleep_coro)
+    retries = 3
+    retry_wait = 5
+
+    async def get_json(url, params=None):
+        raise exceptions.GCPHTTPError('504 Gateway Timeout')
+
+    mocker.patch.object(simple_paging_client, 'get_json', get_json)
+
+    results = await simple_paging_client.list_all_items(
+        conftest.API_BASE_URL, {},
+        retries=retries,
+        retry_wait=retry_wait)
+    assert results == []
+    expected_sleeps = [mocker.call(retry_wait) for _ in range(retries)]
+    sleep_mock.assert_has_calls(expected_sleeps)
+    assert len(caplog.records) == 4
+
+
+@pytest.mark.asyncio
+async def test_list_all_items_partial_failure(
+        mocker, mock_coro, simple_paging_client, caplog):
+    """Some requests fail so partially populated list should be returned."""
+    caplog.set_level(logging.WARNING)
+    sleep_mock, sleep_coro = mock_coro
+    mocker.patch('gordon_janitor_gcp.http_client.asyncio.sleep', sleep_coro)
+    retries = 3
+    retry_wait = 5
+
+    number_of_calls = 0
+
+    async def get_json(url, params=None):
+        nonlocal number_of_calls
+        # One successful call
+        if number_of_calls < 1:
+            number_of_calls += 1
+            return {'data': 'data', 'nextPageToken': 'token'}
+        else:
+            # second call fails despite retries
+            raise exceptions.GCPHTTPError('504 Gateway Timeout')
+
+    mocker.patch.object(simple_paging_client, 'get_json', get_json)
```

Comment:
Is this the way around trying to get `aioresponses` to work the way you want to?

Reply:
Not really, it's because the mixin is easier to test when it's isolated, so it doesn't need to make http requests.